Time Complexity (#3284)
cpinitiative/usaco-guide: content/2_Bronze/Time_Comp.mdx

example is that although binary search on an array and insertion into an ordered
set are both $\mathcal{O}(\log n)$, binary search is noticeably faster.

**Constant factor** is entirely ignored in Big O notation. This is fine most of
the time, but if the time limit is particularly tight, you may receive time
limit exceeded (TLE) with the intended complexity. When this happens, it is
important to keep the constant factor in mind. For example, a piece of code that
iterates through all _ordered_ triplets runs in $\mathcal{O}(n^3)$ time might be
sped up by a factor of $6$ if we only need to iterate through all _unordered_
triplets.

<!-- For example, if our code currently runs in $\mathcal{O}(n^2)$ time, perhaps we can modify our code to make it run in $\mathcal{O}(n^2/32)$ by using a bitset. (Of course, with Big O notation, $\mathcal{O}(n^2) = O(n^2/32)$.) -->
<!-- Bitsets are significantly faster (and so are bitwise operations vs iterating through bits). I don't want to make it sound insignificant -->

For now, don't worry about optimizing constant factors -- just be aware of them.

## Formal Definition of Big O notation

> Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exist
> 
> positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we

Reviewer: ```suggestion
Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exist
positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we
```

say that $f(n) = \mathcal{O}(g(n))$.

> Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.

Reviewer: ```suggestion
Therefore, we *could* say that the time complexity of a linear function, $\mathcal{O}(n)$, is also $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.
However, we usually just write the simplest function out of those that are the most restrictive, which in the case of our linear function above, is $\mathcal{O}(n)$.

> 
> 
> But typically, we write the simplest function out of those that are the most restrictive, thus, $\mathcal{O}(n)$.
> 

Reviewer: ```suggestion
```

Author: i think this is neede cuz it defines what restrictive and simple means, especially on a beginner lvl.

Reviewer: i included it in the previous suggestion

<Optional title="P vs. NP">

P refers to the class of problems that can be solved within polynomial time ($\mathcal{O}(n^2), \mathcal{O}(n^3), \mathcal{O}(n^{100}), \dots$).
NP, short for nondeterministic polynomial time, is the set of problems with solutions that can be verified in polynomial time.

> A common example of a problem in NP is a generalized version of Sudoku, where a solution is easily verifiable in polynomial time, but it is unknown whether a solution is computable in polynomial time. “P vs. NP” is the classic unsolved problem that asks whether every problem that can be verified in polynomial time can also be solved in polynomial time.

Reviewer: > “P vs. NP” is the classic

quotes seem a bit different?

Author: not sure i understand what u mean

Reviewer: `”` vs `"`

example is that although binary search on an array and insertion into an ordered
set are both $\mathcal{O}(\log n)$, binary search is noticeably faster.

**Constant factor** is entirely ignored in Big O notation. This is fine most of
the time, but if the time limit is particularly tight, you may receive time
limit exceeded (TLE) with the intended complexity. When this happens, it is
important to keep the constant factor in mind. For example, a piece of code that
iterates through all _ordered_ triplets runs in $\mathcal{O}(n^3)$ time might be
sped up by a factor of $6$ if we only need to iterate through all _unordered_
triplets.

<!-- For example, if our code currently runs in $\mathcal{O}(n^2)$ time, perhaps we can modify our code to make it run in $\mathcal{O}(n^2/32)$ by using a bitset. (Of course, with Big O notation, $\mathcal{O}(n^2) = O(n^2/32)$.) -->
<!-- Bitsets are significantly faster (and so are bitwise operations vs iterating through bits). I don't want to make it sound insignificant -->

For now, don't worry about optimizing constant factors -- just be aware of them.

## Formal Definition of Big O notation

> Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exists

Reviewer 2: ```suggestion
Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exist

```

positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we
say that $f(n) = \mathcal{O}(g(n))$.

> Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.

Reviewer 2: ```suggestion
Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.

```

> But typically, we write the most restrictive function, thus, $\mathcal{O}(n)$.

Reviewer 2: ```suggestion
But typically, we write the simplest function out of those that are the most restrictive, thus, $\mathcal{O}(n)$.

```

<Optional title="P vs. NP">

P refers to the class of problems that can be solved within polynomial time. (ie. $\mathcal{O}(n^2)$, $\mathcal{O}(n^3)$, $\mathcal{O}(n^100)$...)
NP, known as nondeterministic polynomial, is the set of problems with solutions that can be verified in polynomial time, but not 

> solved in polynomial time.

Reviewer 2: it seems like you're asserting that P != NP here

A common example of NP is the game of Sudoku, where a solution is easily verifiable, but not easily computable (exponentially slower as grid size increases).

> “P = NP” is the classic problem that determines whether a problem that can be verified in polynomial time can also be solved 

Reviewer 2: ```suggestion
“P vs NP” is the classic problem that asks whether every problem that can be verified in polynomial time can also be solved 

```

Reviewer 2: also mention that this is unsolved

in polynomial time.

> If you're interested in learning more about P vs. NP, check out this
> [Youtube video](https://www.youtube.com/watch?v=YX40hbAHx3s). If you're just
> starting out, you may want to skip this video for now.

Reviewer 2: ```suggestion
If you're interested in learning more about P vs. NP, check out this
[Youtube video](https://www.youtube.com/watch?v=YX40hbAHx3s).

```

seems unnecessary to include this line, considering that everything is inside an optional block

positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we
say that $f(n) = \mathcal{O}(g(n))$.

Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.

But typically, we write the simplest function out of those that are the most restrictive, thus, $\mathcal{O}(n)$.

<Optional title="P vs. NP">

P refers to the class of problems that can be solved within polynomial time. (ie. $\mathcal{O}(n^2)$, $\mathcal{O}(n^3)$, $\mathcal{O}(n^100)$...)
NP, known as nondeterministic polynomial, is the set of problems with solutions that can be verified in polynomial time.

A common example of NP is the game of Sudoku, where a solution is easily verifiable, but not easily computable (exponentially slower as grid size increases), though 
there may exist a solution that does solve the solution under polynomial time.

“P vs NP” is the classic problem that asks whether every problem that can be verified in polynomial time can also be solved. This question is currently unsolved. 

> in polynomial time.

Reviewer 2: ^ move this to the correct place

Reviewer 2: I'll fix it
